##import libraries
```{r}
library(ggplot2)
library(DESeq2)
library(pheatmap)
library(forcats)
library(tximport)
library(dplyr)
library(EnhancedVolcano)
library(PCAtools)
library(cowplots)
library(Matrix)
```

set.seed(1)

##samples processing using output files from RSEM with default parameters
```{r}
samples.all <- read.table(file.path("~/RNAseq/samples.txt"))
files.all <- file.path("~/RNAseq", paste0(samples.all$V1, "_withDup.rsem.genes.results"))
txi.rsem.all <- tximport(files.all, type = "rsem")
names(files.all) <- paste0(samples.all$V1)
sampleTreatment.all <- read.csv("/Users/anglo4791/Desktop/Stanford/Microwell/RNAseq/sampleTreatment.csv")
placeholder <- replace(txi.rsem.all$length, txi.rsem.all$length == 0.00, 1.00)
txi.rsem.all$length <- placeholder
rownames(sampleTreatment.all) <- sampleTreatment.all$X
rna_dds <- DESeqDataSetFromTximport(txi.rsem.all, sampleTreatment.all, ~ Treatment)
keep <- rowSums(counts(rna_dds) == 0) <= 5 ##remove genes if >= samples have zero count
dds1 <- rna_dds[keep,]
rna_dds_object <- DESeq(dds1)
rna_rld <- vst(dds1)
rna_rld.assay <- assay(rna_rld)
pcaData <- plotPCA(rna_rld, intgroup=c("Treatment"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

##to plot PCA; FigS13A
ggplot(pcaData, aes(PC1, PC2, color=group)) + geom_point(size=5) +   xlab(paste0("PC1: ",percentVar[1],"% variance")) +   ylab(paste0("PC2: ",percentVar[2],"% variance")) + coord_fixed() + labs(color = "Polarity") + theme_bw() + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), aspect.ratio = 1, axis.text = element_blank(), axis.ticks = element_blank(), axis.title = element_text(size = 20, face = "bold"), legend.title = element_text(size = 15, face = "bold"), legend.text = element_text(size = 12, face = "bold"))
```

##to find significant DE genes
```{r}
deseq_results <- results(rna_dds_object, contrast = c("Treatment", "abnormal", "normal"))
deseq_df <- deseq_results %>% as.data.frame() %>% tibble::rownames_to_column(var = "Gene")
deseq_df %>%dplyr::arrange(dplyr::desc(stat))
signif_genes <- deseq_df %>% dplyr::filter(padj <= 0.1) %>% dplyr::pull(Gene)
signif_rld <- rna_rld.assay %>% as.data.frame() %>% dplyr::filter(rownames(rna_rld.assay) %in% signif_genes)
resultTable <- deseq_results[order(deseq_results$padj),]
resultTable <- subset(resultTable, padj <=0.1)
write.csv(as.data.frame(resultTable), file="~/resultTable_5zero.csv")
```

##generate volcanoplot of differentially expressed genes between organoids with normal versus abnormal polarity; Fig5I
```{r}
up_path <- c("ENSG00000163931_TKT", "ENSG00000184009_ACTG1", "ENSG00000100906_NFKBIA", "ENSG00000188042_ARL4C")
down_path <- c("ENSG00000171316_CHD7", "ENSG00000184368_MAP7D2", "ENSG00000122085_MTERF4", "ENSG00000136731_UGGT1", "ENSG00000204120_GIGYF2", "ENSG00000101966_XIAP", "ENSG00000106541_AGR2", "ENSG00000132341_RAN")
deseq_results$labels <- "NA"
deseq_results$labels[grep("ENSG00000188042_ARL4C", rownames(deseq_results))] <- "ARL4C"
deseq_results$labels[grep("ENSG00000100906_NFKBIA", rownames(deseq_results))] <- "NFKBIA"
keyvals.colour1 <- ifelse(
  rownames(deseq_results) %in% up_path, 'blue',
  ifelse(rownames(deseq_results) %in% down_path, 'red3',
         'grey'))
keyvals.colour1[is.na(keyvals.colour1)] <- 'black'
names(keyvals.colour1)[keyvals.colour1 == 'grey'] <- 'NA'
names(keyvals.colour1)[keyvals.colour1 == 'blue'] <- 'UP_1,2,3,4'
names(keyvals.colour1)[keyvals.colour1 == 'red3'] <- 'DOWN_8,9,10,11'
EnhancedVolcano(deseq_results, lab=deseq_results$labels, selectLab = c("ARL4C", "NFKBIA"), x='log2FoldChange', y='padj', pCutoff = 0.1, col = c('grey', 'grey', 'grey', 'blue'), colAlpha = 1, pointSize = 3, title = "DESeq2 Result", subtitle = "Differential Expression", legendLabels = c("Not significant", "Not significant", "Not significant", "Significant"), legendPosition = 'right', colCustom = keyvals.colour1, labSize = 4, drawConnectors = TRUE, widthConnectors = 0.75) + ylim(0,3) + ylab(bquote(-log[10](FDR))) + xlab(bquote(-log[2](FC))) + theme_bw() + theme(aspect.ratio = 1, panel.grid = element_blank(), legend.title = element_blank(), legend.text = element_text(size = 12), axis.title=element_text(size = 20, face = "bold"), axis.text = element_text(size = 15)) + theme_classic()
```

##single-cell RNA processing; FigS2A
```{r}
scrna.data <- Read10X("~/filtered_gene_bc_matrices/") ##data associated with 10X processing output
Hash_microwell_cells <- scan("~/Hash-seq_microwell_cells.csv", character(), quote = "")
Hash_matrigel_cells <- scan("~/Hash-seq_matrigel_cells.csv", character(), quote = "")

scrna2 <- CreateSeuratObject(scrna.data)
subset_MT_MG <- scrna2[, WhichCells(scrna2, cells = c(Hash_microwell_cells, Hash_matrigel_cells))]
subset_MT_MG_test <- SetIdent(subset_MT_MG, value= "NonHash")
subset_MT_MG_test <- SetIdent(subset_MT_MG_test, cells = Hash_microwell_cells, value= "microwell")
subset_MT_MG_test <- SetIdent(subset_MT_MG_test, cells = Hash_matrigel_cells, value= "matrigel")

scrna3 <- NormalizeData(subset_MT_MG_test)
scrna3 <- FindVariableFeatures(scrna3, do.plot = FALSE, y.cutoff = 0.5)
scrna3 <- ScaleData(scrna3, display.progress = FALSE)
scrna3 <- RunPCA(scrna3, pcs.print = 0)
scrna3 <- FindNeighbors(scrna3, dims = 1:16)
scrna3 <- RunUMAP(scrna3, reduction = "pca", dims = 1:16)

unintegrated_title <- DimPlot(scrna3, cols = c("blue", "gray"), reduction = "umap") + theme(aspect.ratio = 1) + NoLegend() + ggtitle("Bulk vs Microwell")

muc5ac <- FeaturePlot(object = scrna3, features = c("MUC5AC")) + NoLegend() + theme(aspect.ratio = 1)
tff2 <- FeaturePlot(object = scrna3, features = c("TFF2")) + NoLegend() + theme(aspect.ratio = 1)
PGC <- FeaturePlot(object = scrna3, features = c("PGC")) + NoLegend() + theme(aspect.ratio = 1)
gkn2 <- FeaturePlot(object = scrna3, features = c("GKN2")) + NoLegend() + theme(aspect.ratio = 1)
lgr5 <- FeaturePlot(object = scrna3, features = c("LGR5")) + NoLegend() + theme(aspect.ratio = 1)
plot_grid(unintegrated_title, muc5ac, tff2, PGC, gkn2, lgr5)

```
